### ISSUE TITLE
Schnorr - combine 2 signature

### ISSUE THREAD

**Issue by tinguyen1107** [rust-bitcoin/rust-bitcoin#2712]
Hi guys, I'm trying to implement the statecoin protocol, I used the schnorr to handle the multisig between the server and user. We create the signature using this one
``` rust    
    let sighash = sighasher
        .taproot_key_spend_signature_hash(input_index, &prevouts, sighash_type)
        .expect("failed to construct sighash");

    let tweaked: TweakedKeypair = keypair.tap_tweak(&secp, None);
    let msg = Message::from(sighash);

    let signature = secp.sign_schnorr(&msg, &tweaked.to_inner());

    let signature = bitcoin::taproot::Signature {
        sig: signature,
        hash_ty: sighash_type,
    };
```
I wonder that how we can combine 2 sign together and spend the transaction?

**Comment by apoelstra** [comment]
https://9www.vice.com/en/article/wnx8nq/why-you-dont-roll-your-own-crypto

https://galois.com/blog/2021/03/actually-you-are-rolling-your-own-crypto/

**Comment by tinguyen1107** [comment]
Hi @apoelstra, sorry I don't know what you mean? You mean I should not try to implement it on my own? This is for study purpose only, if you know please help me with this :(((

**Comment by apoelstra** [comment]
I do not know anything about the "statecoin protocol". In general it is not possible to combine two Schnorr signatures.

**Comment by storopoli** [comment]
> I do not know anything about the "statecoin protocol". In general it is not possible to combine two Schnorr signatures.

The "In general" is not that helpful.
However, you can combine two Schnorr signatures in something called "Naive Schnorr multi-signatures".
The intuition behind it is that Schnorr sig is "linear", i.e. it does not have a mod inverse in signing like DSA (or ECDSA).

Ref: https://blog.blockstream.com/en-musig-key-aggregation-schnorr-signatures/

**Comment by Kixunil** [comment]
> In general it is not possible to combine two Schnorr signatures.

What?! Isn't that what batch verification does?

**Comment by junderw** [comment]
... I think people are misunderstanding @apoelstra 

I don't think he meant to say that split signature schemes are impossible in Schnorr.

He is saying that using `sign_schnorr` in the way OP does, where two parties are adversarial and can not share the secret nonce without a 2 round scheme like musig2.

OP saying "multisig between the server and user." indicates that the two parties in the multisig should not need to trust each other... so "just add together the s values" is not a viable method of performing an N-of-N multisig scheme.

**Comment by apoelstra** [comment]
>What?! Isn't that what batch verification does?

No, not without stretching the definition of "combine Schnorr signatures" beyond usefulness. Batch verification verifiers multiple Schnorr signatures by doing a large multiexponentiation of nonces and public keys against a randomized linear combination of signature scalar values. I guess you could consider the combination of scalars to be a "combined Schnorr signature" but it exists only ephemerally and does not resemble any part of a signature, nor is it (by itself) evidence that any message(s) were signed by any keys.

Half-aggregation is essentially batch-verification where the combined scalar is not ephemeral but is actually stored in place of the original scalars. It uses publicly-computable randomizers and stores the original nonces. You could call this "combining Schnorr signatures" but the result cannot be verified on-chain today, nor can the original signatures be recovered from it, nor does it have anything to do with multisignatures, so I doubt this is what the OP was asking for.

**Comment by Kixunil** [comment]
What about MuSig2 or full aggregation?

**Comment by apoelstra** [comment]
Neither are "in general" nor do they start with Schnorr signatures. They start with partial signatures and require specific protocols to produce.

**Comment by Kixunil** [comment]
@junderw I only noticed your comment now. That's what I though was going on but the way @apoelstra phrased it made me think it's actually *mathematically impossible* to combine the signatures, not just horribly unsecure.

So IIUC @tinguyen1107 don't do that if you want to keep your money. You need to wait for MuSig2 support. There's already a PR to upstream MuSig2 to `libsecp256k1`, then we'll need to add the API here. I might actually be able to do this part as a paid job, we'll see.
